src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -4405,6 +4405,197 @@ class StubGenerator: public StubCodeGenerator {
+    StubCodeMark mark(this, stub_id);
+    address start = __ pc();
+    __ enter();
+
+    int i, j;
+    const Register state = c_rarg0;
+    const Register keystream = c_rarg1;
+    const Register loopCtr = r10;
+    const Register tmpAddr = r11;
+    const FloatRegister ctrAddOverlay = v28;
+    const FloatRegister lrot8Tbl = v29;
+
+    // Organize SIMD registers in an array that facilitates
+    // putting repetitive opcodes into loop structures.  It is
+    // important that each grouping of 4 registers is monotonically
+    // increasing to support the requirements of multi-register
+    // instructions (e.g. ld4r, st4, etc.)
+    const FloatRegister workSt[16] = {
+         v4,  v5,  v6,  v7, v16, v17, v18, v19,
+        v20, v21, v22, v23, v24, v25, v26, v27
+    };
+
+    // Pull in constant data.  The first 16 bytes are the add overlay
+    // which is applied to the vector holding the counter (state[12]).
+    // The second 16 bytes is the index register for the 8-bit left
+    // rotation tbl instruction.
+    __ adr(tmpAddr, L_cc20_const);
+    __ ldpq(ctrAddOverlay, lrot8Tbl, Address(tmpAddr));
+
+    // Load from memory and interlace across 16 SIMD registers,
+    // With each word from memory being broadcast to all lanes of
+    // each successive SIMD register.
+    //      Addr(0) -> All lanes in workSt[i]
+    //      Addr(4) -> All lanes workSt[i + 1], etc.
+    __ mov(tmpAddr, state);
+    for (i = 0; i < 16; i += 4) {
+      __ ld4r(workSt[i], workSt[i + 1], workSt[i + 2], workSt[i + 3], __ T4S,
+          __ post(tmpAddr, 16));
+    }
+    __ addv(workSt[12], __ T4S, workSt[12], ctrAddOverlay); // Add ctr overlay
+
+    // Before entering the loop, create 5 4-register arrays.  These
+    // will hold the 4 registers that represent the a/b/c/d fields
+    // in the quarter round operation.  For instance the "b" field
+    // for the first 4 quarter round operations is the set of v16/v17/v18/v19,
+    // but in the second 4 quarter rounds it gets adjusted to v17/v18/v19/v16
+    // since it is part of a diagonal organization.  The aSet and scratch
+    // register sets are defined at declaration time because they do not change
+    // organization at any point during the 20-round processing.
+    FloatRegister aSet[4] = { v4, v5, v6, v7 };
+    FloatRegister bSet[4];
+    FloatRegister cSet[4];
+    FloatRegister dSet[4];
+    FloatRegister scratch[4] = { v0, v1, v2, v3 };
+
+    // Set up the 10 iteration loop and perform all 8 quarter round ops
+    __ mov(loopCtr, 10);
+    __ BIND(L_twoRounds);
+
+    // Set to columnar organization and do the following 4 quarter-rounds:
+    // QUARTERROUND(0, 4, 8, 12)
+    // QUARTERROUND(1, 5, 9, 13)
+    // QUARTERROUND(2, 6, 10, 14)
+    // QUARTERROUND(3, 7, 11, 15)
+    bSet[0] = v16; bSet[1] = v17; bSet[2] = v18; bSet[3] = v19;
+    cSet[0] = v20; cSet[1] = v21; cSet[2] = v22; cSet[3] = v23;
+    dSet[0] = v24; dSet[1] = v25; dSet[2] = v26; dSet[3] = v27;
```

Comment:
```suggestion
    bSet[0] = workSt[4];  bSet[1] = workSt[5];  bSet[2] = workSt[6];  bSet[3] = workSt[7];
    cSet[0] = workSt[8];  cSet[1] = workSt[9];  cSet[2] = workSt[10]; cSet[3] = workSt[11];
    dSet[0] = workSt[12]; dSet[1] = workSt[13]; dSet[2] = workSt[14]; dSet[3] = workSt[15];
```
How about something like this? The mapping from index to register, and from specification to implementation, is easier for this reviewer to understand. 

or maybe (better?) define a function, such that:

```
regs_for_quarter_round(bSet, workSt, 4, 5, 6, 7);
regs_for_quarter_round(cSet, workSt, 8, 9, 10, 11);
regs_for_quarter_round(dSet, workSt, 12, 13, 14, 15);

```

Reply:
I like either approach, I'll try the function route.  Either way it definitely helps make things more clear.  Putting this in terms of the workSt indicies maps more closely to how things are described in the RFC.  Good call!